File: go-spider/internal/security/osec_evidence.go
package security

import (
	"bytes"
	"crypto/sha256"
	"encoding/binary"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"math/big"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"unicode/utf8"
)

const (
	aggregateV1MagicValue         = "OSEC-CASE-AGGREGATE-V1\x00"
	categoryAggregateV1MagicValue = "OSEC-CASE-CATEGORY-AGGREGATE-V1\x00"
	maxU32                        = uint64(0xFFFFFFFF)
)

var (
	caseIDPattern       = regexp.MustCompile(`^[a-z][a-z0-9-]{0,63}$`)
	categoryNamePattern = regexp.MustCompile(`^[a-z][a-z0-9_]{0,63}$`)
)

// ValueKind is the explicit OSEC typed AST kind.
type ValueKind int

const (
	NullKind ValueKind = iota
	BoolKind
	IntegerKind
	StringKind
	ArrayKind
	ObjectKind
	NonIntegerKind
)

// ObjectMember preserves decoded object key order and raw member count.
type ObjectMember struct {
	Key   string
	Value Value
}

// Value is the cross-language OSEC typed AST value.
type Value struct {
	Kind          ValueKind
	Bool          bool
	Integer       *big.Int
	IntegerLexeme string
	String        string
	Array         []Value
	Object        []ObjectMember
	NonInteger    string
}

// EvidenceLimits carries frozen OSEC-EVIDENCE-LIMITS-V1 values.
type EvidenceLimits struct {
	FileSize      int
	NestingDepth  int
	IntegerDigits int
	ArrayLength   int
	ObjectMembers int
	StringLength  int
	CaseCount     int
}

// DefaultEvidenceLimits returns the frozen default limits.
func DefaultEvidenceLimits() EvidenceLimits {
	return EvidenceLimits{
		FileSize:      16_777_216,
		NestingDepth:  128,
		IntegerDigits: 4096,
		ArrayLength:   10_000,
		ObjectMembers: 1_000,
		StringLength:  1_048_576,
		CaseCount:     10_000,
	}
}

// EvidenceError is the stable machine-code error.
type EvidenceError struct {
	Code    string
	Message string
}

func (e *EvidenceError) Error() string {
	return fmt.Sprintf("OSEC error %s: %s", e.Code, e.Message)
}

func codeError(code string) error {
	return &EvidenceError{Code: code, Message: code}
}

// CaseDigest is an id plus a 32-byte digest.
type CaseDigest struct {
	ID     string
	Digest []byte
}

// CanonicalCaseResult returns canonical bytes and digest.
type CanonicalCaseResult struct {
	Canonical []byte
	SHA256    string
}

func narrowSurrogatePrecheck(raw []byte) error {
	inString := false
	i := 0
	n := len(raw)
	for i < n {
		b := raw[i]
		if !inString {
			if b == '"' {
				inString = true
			}
			i++
			continue
		}
		if b == '\\' {
			if i+1 >= n {
				return fmt.Errorf("truncated escape")
			}
			next := raw[i+1]
			if next == 'u' {
				if i+6 > n {
					return fmt.Errorf("truncated unicode escape")
				}
				first, err := strconv.ParseUint(string(raw[i+2:i+6]), 16, 16)
				if err != nil {
					return err
				}
				if first >= 0xD800 && first <= 0xDBFF {
					if i+12 > n || raw[i+6] != '\\' || raw[i+7] != 'u' {
						return fmt.Errorf("lone high surrogate")
					}
					second, err := strconv.ParseUint(string(raw[i+8:i+12]), 16, 16)
					if err != nil || second < 0xDC00 || second > 0xDFFF {
						return fmt.Errorf("invalid surrogate pair")
					}
					i += 12
					continue
				}
				if first >= 0xDC00 && first <= 0xDFFF {
					return fmt.Errorf("lone low surrogate")
				}
				i += 6
				continue
			}
			i += 2
			continue
		}
		if b == '"' {
			inString = false
		}
		i++
	}
	return nil
}

func decodeValue(dec *json.Decoder) (Value, string, error) {
	tok, err := dec.Token()
	if err != nil {
		return Value{}, "", err
	}
	switch t := tok.(type) {
	case nil:
		return Value{Kind: NullKind}, "", nil
	case bool:
		return Value{Kind: BoolKind, Bool: t}, "", nil
	case string:
		return Value{Kind: StringKind, String: t}, "", nil
	case json.Number:
		lex := string(t)
		if strings.ContainsAny(lex, ".eE") {
			return Value{Kind: NonIntegerKind, NonInteger: lex}, "", nil
		}
		return Value{Kind: IntegerKind, IntegerLexeme: lex}, "", nil
	case json.Delim:
		switch t {
		case '[':
			var arr []Value
			dup := ""
			for dec.More() {
				item, nestedDup, err := decodeValue(dec)
				if err != nil {
					return Value{}, "", err
				}
				if dup == "" && nestedDup != "" {
					dup = nestedDup
				}
				arr = append(arr, item)
			}
			if _, err := dec.Token(); err != nil {
				return Value{}, "", err
			}
			return Value{Kind: ArrayKind, Array: arr}, dup, nil
		case '{':
			var members []ObjectMember
			seen := map[string]bool{}
			dup := ""
			for dec.More() {
				keyTok, err := dec.Token()
				if err != nil {
					return Value{}, "", err
				}
				key, ok := keyTok.(string)
				if !ok {
					return Value{}, "", fmt.Errorf("object key is not string")
				}
				if seen[key] && dup == "" {
					dup = key
				}
				seen[key] = true
				item, nestedDup, err := decodeValue(dec)
				if err != nil {
					return Value{}, "", err
				}
				if dup == "" && nestedDup != "" {
					dup = nestedDup
				}
				members = append(members, ObjectMember{Key: key, Value: item})
			}
			if _, err := dec.Token(); err != nil {
				return Value{}, "", err
			}
			return Value{Kind: ObjectKind, Object: members}, dup, nil
		}
	}
	return Value{}, "", fmt.Errorf("unknown JSON token")
}

// StrictDecode implements OSEC-STRICT-JSON-DECODE-V1.
func StrictDecode(raw []byte, limits EvidenceLimits) (Value, error) {
	if len(raw) > limits.FileSize {
		return Value{}, codeError("evidence_limit_file_size")
	}
	if bytes.HasPrefix(raw, []byte{0xEF, 0xBB, 0xBF}) {
		return Value{}, codeError("strict_decode_bom")
	}
	if !utf8.Valid(raw) {
		return Value{}, codeError("strict_decode_invalid_utf8")
	}
	trimmed := bytes.TrimLeft(raw, " \t\r\n")
	if len(trimmed) >= 2 && trimmed[0] == '0' && trimmed[1] >= '0' && trimmed[1] <= '9' {
		return Value{}, codeError("strict_decode_invalid_json")
	}
	if len(trimmed) >= 3 && trimmed[0] == '-' && trimmed[1] == '0' && trimmed[2] >= '0' && trimmed[2] <= '9' {
		return Value{}, codeError("strict_decode_invalid_json")
	}
	if numberPrefixEndsTruncated(trimmed) {
		return Value{}, codeError("strict_decode_invalid_json")
	}

	surrogateErr := narrowSurrogatePrecheck(raw)
	dec := json.NewDecoder(bytes.NewReader(raw))
	dec.UseNumber()
	value, dup, err := decodeValue(dec)
	if err != nil {
		return Value{}, codeError("strict_decode_invalid_json")
	}
	if _, err := dec.Token(); err != io.EOF {
		return Value{}, codeError("strict_decode_trailing_data")
	}
	if surrogateErr != nil {
		return Value{}, codeError("strict_decode_lone_surrogate")
	}
	if dup != "" {
		return Value{}, codeError("strict_decode_duplicate_key")
	}
	if err := applyLimits(&value, 1, limits); err != nil {
		return Value{}, err
	}
	return value, nil
}

func numberPrefixEndsTruncated(trimmed []byte) bool {
	i := 0
	n := len(trimmed)
	if i < n && trimmed[i] == '-' {
		i++
	}
	if i >= n {
		return false
	}
	if trimmed[i] == '0' {
		i++
	} else if trimmed[i] >= '1' && trimmed[i] <= '9' {
		for i < n && trimmed[i] >= '0' && trimmed[i] <= '9' {
			i++
		}
	} else {
		return false
	}
	if i < n && trimmed[i] == '.' {
		i++
		digits := 0
		for i < n && trimmed[i] >= '0' && trimmed[i] <= '9' {
			i++
			digits++
		}
		if digits == 0 {
			return true
		}
	}
	if i < n && (trimmed[i] == 'e' || trimmed[i] == 'E') {
		i++
		if i < n && (trimmed[i] == '+' || trimmed[i] == '-') {
			i++
		}
		digits := 0
		for i < n && trimmed[i] >= '0' && trimmed[i] <= '9' {
			i++
			digits++
		}
		if digits == 0 {
			return true
		}
	}
	return false
}

func applyLimits(value *Value, depth int, limits EvidenceLimits) error {
	switch value.Kind {
	case ObjectKind:
		if depth > limits.NestingDepth {
			return codeError("evidence_limit_nesting_depth")
		}
		if len(value.Object) > limits.ObjectMembers {
			return codeError("evidence_limit_object_members")
		}
		sort.SliceStable(value.Object, func(i, j int) bool {
			return bytes.Compare([]byte(value.Object[i].Key), []byte(value.Object[j].Key)) < 0
		})
		for i := range value.Object {
			if len([]byte(value.Object[i].Key)) > limits.StringLength {
				return codeError("evidence_limit_string_length")
			}
			if err := applyLimits(&value.Object[i].Value, depth+1, limits); err != nil {
				return err
			}
		}
	case ArrayKind:
		if depth > limits.NestingDepth {
			return codeError("evidence_limit_nesting_depth")
		}
		if len(value.Array) > limits.ArrayLength {
			return codeError("evidence_limit_array_length")
		}
		for i := range value.Array {
			if err := applyLimits(&value.Array[i], depth+1, limits); err != nil {
				return err
			}
		}
	case StringKind:
		if len([]byte(value.String)) > limits.StringLength {
			return codeError("evidence_limit_string_length")
		}
	case IntegerKind:
		digits := strings.TrimPrefix(value.IntegerLexeme, "-")
		if len(digits) > limits.IntegerDigits {
			return codeError("evidence_limit_integer_digits")
		}
		parsed, ok := new(big.Int).SetString(value.IntegerLexeme, 10)
		if !ok {
			return codeError("canonical_invalid_value_type")
		}
		value.Integer = parsed
	}
	return nil
}

func validateCanonicalValue(value Value) error {
	switch value.Kind {
	case NullKind, BoolKind:
		return nil
	case IntegerKind:
		if value.Integer == nil {
			return codeError("canonical_invalid_value_type")
		}
		return nil
	case NonIntegerKind:
		return codeError("canonical_non_integer_number")
	case StringKind:
		if !utf8.ValidString(value.String) {
			return codeError("canonical_invalid_unicode")
		}
		return nil
	case ArrayKind:
		for _, item := range value.Array {
			if err := validateCanonicalValue(item); err != nil {
				return err
			}
		}
		return nil
	case ObjectKind:
		for _, member := range value.Object {
			if !utf8.ValidString(member.Key) {
				return codeError("canonical_invalid_unicode")
			}
			if err := validateCanonicalValue(member.Value); err != nil {
				return err
			}
		}
		return nil
	default:
		return codeError("canonical_invalid_value_type")
	}
}

func appendCanonicalString(buf *bytes.Buffer, value string) error {
	if !utf8.ValidString(value) {
		return codeError("canonical_invalid_unicode")
	}
	buf.WriteByte('"')
	for _, r := range value {
		switch r {
		case '"':
			buf.WriteString(`\"`)
		case '\\':
			buf.WriteString(`\\`)
		case '\b':
			buf.WriteString(`\b`)
		case '\t':
			buf.WriteString(`\t`)
		case '\n':
			buf.WriteString(`\n`)
		case '\f':
			buf.WriteString(`\f`)
		case '\r':
			buf.WriteString(`\r`)
		default:
			if r < 0x20 {
				_, _ = fmt.Fprintf(buf, `\u%04x`, r)
			} else {
				buf.WriteRune(r)
			}
		}
	}
	buf.WriteByte('"')
	return nil
}

func encodeCanonicalValue(buf *bytes.Buffer, value Value) error {
	switch value.Kind {
	case NullKind:
		buf.WriteString("null")
	case BoolKind:
		if value.Bool {
			buf.WriteString("true")
		} else {
			buf.WriteString("false")
		}
	case IntegerKind:
		buf.WriteString(value.Integer.String())
	case StringKind:
		return appendCanonicalString(buf, value.String)
	case ArrayKind:
		buf.WriteByte('[')
		for i, item := range value.Array {
			if i > 0 {
				buf.WriteByte(',')
			}
			if err := encodeCanonicalValue(buf, item); err != nil {
				return err
			}
		}
		buf.WriteByte(']')
	case ObjectKind:
		sorted := append([]ObjectMember(nil), value.Object...)
		sort.SliceStable(sorted, func(i, j int) bool {
			return bytes.Compare([]byte(sorted[i].Key), []byte(sorted[j].Key)) < 0
		})
		buf.WriteByte('{')
		for i, member := range sorted {
			if i > 0 {
				buf.WriteByte(',')
			}
			if err := appendCanonicalString(buf, member.Key); err != nil {
				return err
			}
			buf.WriteByte(':')
			if err := encodeCanonicalValue(buf, member.Value); err != nil {
				return err
			}
		}
		buf.WriteByte('}')
	default:
		return codeError("canonical_invalid_value_type")
	}
	return nil
}

// CanonicalJSON implements OSEC-CANONICAL-JSON-V1.
func CanonicalJSON(value Value) ([]byte, error) {
	if err := validateCanonicalValue(value); err != nil {
		return nil, err
	}
	var buf bytes.Buffer
	if err := encodeCanonicalValue(&buf, value); err != nil {
		return nil, err
	}
	return buf.Bytes(), nil
}

// CanonicalCase implements OSEC-CANONICAL-CASE-V1.
func CanonicalCase(value Value) (CanonicalCaseResult, error) {
	if value.Kind != ObjectKind {
		return CanonicalCaseResult{}, codeError("canonical_root_not_object")
	}
	var idValue *Value
	for i := range value.Object {
		if value.Object[i].Key == "id" {
			idValue = &value.Object[i].Value
			break
		}
	}
	if idValue == nil {
		return CanonicalCaseResult{}, codeError("canonical_missing_id")
	}
	if idValue.Kind != StringKind || !caseIDPattern.MatchString(idValue.String) {
		return CanonicalCaseResult{}, codeError("canonical_invalid_id")
	}
	canonical, err := CanonicalJSON(value)
	if err != nil {
		return CanonicalCaseResult{}, err
	}
	sum := sha256.Sum256(canonical)
	return CanonicalCaseResult{Canonical: canonical, SHA256: hex.EncodeToString(sum[:])}, nil
}

// ValidateCaseDataset applies the dataset-level case_count limit before processing.
func ValidateCaseDataset(cases []Value, limits EvidenceLimits) (int, error) {
	count := len(cases)
	if count > limits.CaseCount {
		return count, codeError("evidence_limit_case_count")
	}
	return count, nil
}

func checkedU32(value uint64, code string) (uint32, error) {
	if value > maxU32 {
		return 0, codeError(code)
	}
	return uint32(value), nil
}

func aggregateRecords(magic string, prefix []byte, digests []CaseDigest) (string, error) {
	if len(digests) == 0 {
		return "", codeError("aggregate_empty_set")
	}
	seen := map[string]bool{}
	type record struct {
		id     []byte
		digest []byte
	}
	records := make([]record, 0, len(digests))
	for _, item := range digests {
		if seen[item.ID] {
			return "", codeError("aggregate_duplicate_id")
		}
		seen[item.ID] = true
		if len(item.Digest) != 32 {
			return "", codeError("aggregate_invalid_digest_length")
		}
		records = append(records, record{id: []byte(item.ID), digest: item.Digest})
	}
	sort.Slice(records, func(i, j int) bool {
		return bytes.Compare(records[i].id, records[j].id) < 0
	})
	count, err := checkedU32(uint64(len(records)), "aggregate_count_overflow")
	if err != nil {
		return "", err
	}
	stream := []byte(magic)
	stream = append(stream, prefix...)
	var countBytes [4]byte
	binary.BigEndian.PutUint32(countBytes[:], count)
	stream = append(stream, countBytes[:]...)
	for _, rec := range records {
		length, err := checkedU32(uint64(len(rec.id)), "aggregate_length_overflow")
		if err != nil {
			return "", err
		}
		var lengthBytes [4]byte
		binary.BigEndian.PutUint32(lengthBytes[:], length)
		stream = append(stream, lengthBytes[:]...)
		stream = append(stream, rec.id...)
		stream = append(stream, rec.digest...)
	}
	sum := sha256.Sum256(stream)
	return hex.EncodeToString(sum[:]), nil
}

// AggregateV1 implements OSEC-CASE-AGGREGATE-V1.
func AggregateV1(digests []CaseDigest, algorithm string) (string, error) {
	if algorithm != aggregateV1MagicValue[:len(aggregateV1MagicValue)-1] {
		return "", codeError("aggregate_unknown_algorithm")
	}
	return aggregateRecords(aggregateV1MagicValue, nil, digests)
}

// CategoryAggregateV1 implements OSEC-CASE-CATEGORY-AGGREGATE-V1.
func CategoryAggregateV1(categoryName string, digests []CaseDigest) (string, error) {
	if !categoryNamePattern.MatchString(categoryName) {
		return "", codeError("manifest_invalid_category")
	}
	encoded := []byte(categoryName)
	length, err := checkedU32(uint64(len(encoded)), "aggregate_length_overflow")
	if err != nil {
		return "", err
	}
	var lengthBytes [4]byte
	binary.BigEndian.PutUint32(lengthBytes[:], length)
	prefix := append(lengthBytes[:], encoded...)
	return aggregateRecords(categoryAggregateV1MagicValue, prefix, digests)
}

File: crawler/security/osec_evidence.py
"""OSEC Evidence Profile V2 candidate library.

This module is intentionally not imported by any production entrypoint.
"""

from __future__ import annotations

import hashlib
import json
import re
import struct
from dataclasses import dataclass
from typing import Any, Sequence

_STRICT_ERROR_CODES = frozenset(
    {
        "strict_decode_invalid_json",
        "strict_decode_invalid_utf8",
        "strict_decode_bom",
        "strict_decode_duplicate_key",
        "strict_decode_lone_surrogate",
        "strict_decode_trailing_data",
    }
)
_LIMIT_ERROR_CODES = frozenset(
    {
        "evidence_limit_file_size",
        "evidence_limit_nesting_depth",
        "evidence_limit_integer_digits",
        "evidence_limit_array_length",
        "evidence_limit_object_members",
        "evidence_limit_string_length",
        "evidence_limit_case_count",
    }
)
_CANONICAL_ERROR_CODES = frozenset(
    {
        "canonical_invalid_value_type",
        "canonical_non_integer_number",
        "canonical_invalid_unicode",
        "canonical_root_not_object",
        "canonical_missing_id",
        "canonical_invalid_id",
    }
)
_AGGREGATE_ERROR_CODES = frozenset(
    {
        "aggregate_empty_set",
        "aggregate_duplicate_id",
        "aggregate_invalid_digest_length",
        "aggregate_unknown_algorithm",
        "aggregate_count_overflow",
        "aggregate_length_overflow",
    }
)
_MANIFEST_ERROR_CODES = frozenset({"manifest_invalid_category"})

FROZEN_ERROR_CODES = frozenset(
    _STRICT_ERROR_CODES
    | _LIMIT_ERROR_CODES
    | _CANONICAL_ERROR_CODES
    | _AGGREGATE_ERROR_CODES
    | _MANIFEST_ERROR_CODES
)

AGGREGATE_V1_MAGIC = b"OSEC-CASE-AGGREGATE-V1\x00"
CATEGORY_AGGREGATE_V1_MAGIC = b"OSEC-CASE-CATEGORY-AGGREGATE-V1\x00"
CASE_ID_RE = re.compile(r"^[a-z][a-z0-9-]{0,63}$")
CATEGORY_NAME_RE = re.compile(r"^[a-z][a-z0-9_]{0,63}$")
U32_MAX = 0xFFFFFFFF


class OSECError(Exception):
    """Stable machine-code error without raw input."""

    def __init__(self, code: str, message: str | None = None) -> None:
        self.code = code
        self.message = message if message is not None else code
        super().__init__(f"OSEC error {code}")

    def __str__(self) -> str:
        return f"OSEC error {self.code}: {self.message}"


def _err(code: str) -> OSECError:
    return OSECError(code)


@dataclass(frozen=True)
class EvidenceLimits:
    file_size: int = 16_777_216
    nesting_depth: int = 128
    integer_digits: int = 4096
    array_length: int = 10_000
    object_members: int = 1_000
    string_length: int = 1_048_576
    case_count: int = 10_000


DEFAULT_LIMITS = EvidenceLimits()


class IntegerLexeme(str):
    """Raw integer lexeme used only during parsing."""


class NonIntegerNumber(str):
    """Raw legal JSON number with a decimal point or exponent."""


@dataclass(frozen=True)
class CaseDigest:
    id: str
    digest: bytes


@dataclass(frozen=True)
class CanonicalCaseResult:
    canonical: bytes
    sha256: str


def _normalize_string(value: str) -> str:
    try:
        return value.encode("utf-8", "surrogatepass").decode("utf-8")
    except UnicodeDecodeError as exc:
        raise _err("strict_decode_lone_surrogate") from exc


def _normalize_ast(value: Any) -> Any:
    if isinstance(value, (IntegerLexeme, NonIntegerNumber)):
        return value
    if isinstance(value, str):
        return _normalize_string(value)
    if isinstance(value, list):
        return [_normalize_ast(item) for item in value]
    if isinstance(value, dict):
        return {_normalize_string(key): _normalize_ast(item) for key, item in value.items()}
    return value


def _apply_limits(value: Any, depth: int, raw_objects: list[list[tuple[str, Any]]], index: list[int], limits: EvidenceLimits) -> Any:
    if isinstance(value, IntegerLexeme):
        digits = len(value.lstrip("-"))
        if digits > limits.integer_digits:
            raise _err("evidence_limit_integer_digits")
        return int(value)
    if isinstance(value, NonIntegerNumber):
        return value
    if isinstance(value, dict):
        if depth > limits.nesting_depth:
            raise _err("evidence_limit_nesting_depth")
        pairs = raw_objects[index[0]]
        index[0] += 1
        if len(pairs) > limits.object_members:
            raise _err("evidence_limit_object_members")
        result: dict[str, Any] = {}
        for key, item in sorted(value.items(), key=lambda kv: kv[0].encode("utf-8")):
            if len(key.encode("utf-8")) > limits.string_length:
                raise _err("evidence_limit_string_length")
            result[key] = _apply_limits(item, depth + 1, raw_objects, index, limits)
        return result
    if isinstance(value, list):
        if depth > limits.nesting_depth:
            raise _err("evidence_limit_nesting_depth")
        if len(value) > limits.array_length:
            raise _err("evidence_limit_array_length")
        return [_apply_limits(item, depth + 1, raw_objects, index, limits) for item in value]
    if isinstance(value, str):
        if len(value.encode("utf-8")) > limits.string_length:
            raise _err("evidence_limit_string_length")
        return value
    return value


def strict_decode(raw: bytes, limits: EvidenceLimits = DEFAULT_LIMITS) -> Any:
    if not isinstance(raw, bytes):
        raise TypeError("raw must be bytes")
    if len(raw) > limits.file_size:
        raise _err("evidence_limit_file_size")
    if raw.startswith(b"\xef\xbb\xbf"):
        raise _err("strict_decode_bom")
    try:
        text = raw.decode("utf-8")
    except UnicodeDecodeError as exc:
        raise _err("strict_decode_invalid_utf8") from exc

    stripped = text.lstrip()
    if re.match(r"^-?0[0-9]", stripped):
        raise _err("strict_decode_invalid_json")
    number_prefix = re.match(r"-?(?:0|[1-9][0-9]*)(?:\.[0-9]+)?(?:[eE][+-]?[0-9]*)?", stripped)
    if number_prefix and re.search(r"[eE][+-]?$", number_prefix.group(0)):
        raise _err("strict_decode_invalid_json")

    raw_objects: list[list[tuple[str, Any]]] = []

    def record_pairs(pairs):
        raw_objects.append(list(pairs))
        return dict(pairs)

    def reject_constant(_value: str) -> Any:
        raise ValueError("non-standard JSON constant")

    decoder = json.JSONDecoder(
        parse_int=IntegerLexeme,
        parse_float=NonIntegerNumber,
        parse_constant=reject_constant,
        object_pairs_hook=record_pairs,
    )
    try:
        value, end = decoder.raw_decode(text)
    except (ValueError, json.JSONDecodeError) as exc:
        raise _err("strict_decode_invalid_json") from exc
    if text[end:].strip():
        raise _err("strict_decode_trailing_data")

    value = _normalize_ast(value)
    for pairs in raw_objects:
        seen = set()
        for key, _item in pairs:
            normalized = _normalize_string(key)
            if normalized in seen:
                raise _err("strict_decode_duplicate_key")
            seen.add(normalized)
    return _apply_limits(value, 1, raw_objects, [0], limits)


def _validate_canonical_value(value: Any, stack: list[int]) -> Any:
    if value is None or isinstance(value, bool):
        return value
    if isinstance(value, int):
        return value
    if isinstance(value, NonIntegerNumber):
        raise _err("canonical_non_integer_number")
    if isinstance(value, str):
        try:
            return value.encode("utf-8", "surrogatepass").decode("utf-8")
        except UnicodeDecodeError as exc:
            raise _err("canonical_invalid_unicode") from exc
    if isinstance(value, list):
        marker = id(value)
        if marker in stack:
            raise _err("canonical_invalid_value_type")
        stack.append(marker)
        result = [_validate_canonical_value(item, stack) for item in value]
        stack.pop()
        return result
    if isinstance(value, dict):
        marker = id(value)
        if marker in stack:
            raise _err("canonical_invalid_value_type")
        stack.append(marker)
        result = {}
        for key, item in value.items():
            if not isinstance(key, str):
                raise _err("canonical_invalid_value_type")
            normalized_key = _validate_canonical_value(key, stack)
            result[normalized_key] = _validate_canonical_value(item, stack)
        stack.pop()
        return result
    raise _err("canonical_invalid_value_type")


def _encode_string(value: str) -> bytes:
    out: list[str] = ['"']
    for char in value:
        code = ord(char)
        if char == '"':
            out.append('\\"')
        elif char == "\\":
            out.append("\\\\")
        elif char == "\b":
            out.append("\\b")
        elif char == "\t":
            out.append("\\t")
        elif char == "\n":
            out.append("\\n")
        elif char == "\f":
            out.append("\\f")
        elif char == "\r":
            out.append("\\r")
        elif code < 0x20:
            out.append("\\u%04x" % code)
        else:
            out.append(char)
    out.append('"')
    return "".join(out).encode("utf-8")


def _encode_canonical(value: Any) -> bytes:
    if value is None:
        return b"null"
    if value is True:
        return b"true"
    if value is False:
        return b"false"
    if isinstance(value, int):
        return str(value).encode("ascii")
    if isinstance(value, str):
        return _encode_string(value)
    if isinstance(value, list):
        return b"[" + b",".join(_encode_canonical(item) for item in value) + b"]"
    if isinstance(value, dict):
        items = sorted(value.items(), key=lambda kv: kv[0].encode("utf-8"))
        return b"{" + b",".join(_encode_string(key) + b":" + _encode_canonical(item) for key, item in items) + b"}"
    raise _err("canonical_invalid_value_type")


def canonical_json(value: Any) -> bytes:
    validated = _validate_canonical_value(value, [])
    return _encode_canonical(validated)


def canonical_case(value: Any) -> CanonicalCaseResult:
    if not isinstance(value, dict):
        raise _err("canonical_root_not_object")
    if "id" not in value:
        raise _err("canonical_missing_id")
    case_id = value["id"]
    if not isinstance(case_id, str) or not CASE_ID_RE.match(case_id):
        raise _err("canonical_invalid_id")
    canonical = canonical_json(value)
    return CanonicalCaseResult(canonical=canonical, sha256=hashlib.sha256(canonical).hexdigest())


def validate_case_dataset(cases: Sequence[Any], limits: EvidenceLimits = DEFAULT_LIMITS) -> int:
    count = len(cases)
    if count > limits.case_count:
        raise _err("evidence_limit_case_count")
    return count


def _checked_u32(value: int, code: str) -> int:
    if value < 0 or value > U32_MAX:
        raise _err(code)
    return value


def _aggregate_records(magic: bytes, prefix: bytes | None, digests: Sequence[CaseDigest]) -> str:
    if not digests:
        raise _err("aggregate_empty_set")
    seen = set()
    records = []
    for item in digests:
        if item.id in seen:
            raise _err("aggregate_duplicate_id")
        seen.add(item.id)
        if len(item.digest) != 32:
            raise _err("aggregate_invalid_digest_length")
        records.append((item.id.encode("utf-8"), item.digest))
    records.sort(key=lambda record: record[0])
    _checked_u32(len(records), "aggregate_count_overflow")
    stream = bytearray(magic)
    if prefix is not None:
        stream.extend(prefix)
    stream.extend(struct.pack(">I", _checked_u32(len(records), "aggregate_count_overflow")))
    for case_id, digest in records:
        stream.extend(struct.pack(">I", _checked_u32(len(case_id), "aggregate_length_overflow")))
        stream.extend(case_id)
        stream.extend(digest)
    return hashlib.sha256(bytes(stream)).hexdigest()


def aggregate_v1(digests: Sequence[CaseDigest], algorithm: str = "OSEC-CASE-AGGREGATE-V1") -> str:
    if algorithm != "OSEC-CASE-AGGREGATE-V1":
        raise _err("aggregate_unknown_algorithm")
    return _aggregate_records(AGGREGATE_V1_MAGIC, None, digests)


def category_aggregate_v1(category_name: str, digests: Sequence[CaseDigest]) -> str:
    if not CATEGORY_NAME_RE.match(category_name):
        raise _err("manifest_invalid_category")
    encoded = category_name.encode("utf-8")
    prefix = struct.pack(">I", _checked_u32(len(encoded), "aggregate_length_overflow")) + encoded
    return _aggregate_records(CATEGORY_AGGREGATE_V1_MAGIC, prefix, digests)
